packages/core/src/render3/framework-injector-profiler.ts

```diff
@@ -0,0 +1,121 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {InjectionToken} from '../di/injection_token';
+import {Injector} from '../di/injector';
+import {EnvironmentInjector} from '../di/r3_injector';
+import {Type} from '../interface/type';
+
+import {isStandalone} from './definition';
+import {DebugNodeInjector} from './di';
+import {getDebugInjectContext, InjectedService, InjectorProfilerEvent, InjectorProfilerEventType, ProviderRecord, setInjectorProfiler} from './injector-profiler';
+import {LView} from './interfaces/view';
+
+export const frameworkDIDebugData = {
+  injectorToInstantiatedTokenToDependencies:
+      new WeakMap<Injector|LView, WeakMap<Type<unknown>, InjectedService[]>>(),
+  instanceToInjector: new WeakMap<object, Injector>(),
+  injectorToProviders: new WeakMap<Injector|LView, ProviderRecord[]>(),
+  standaloneInjectorToComponent: new WeakMap<Injector, Type<unknown>>()
+};
+
+export function setupFrameworkInjectorProfiler(): void {
+  frameworkDIDebugData.injectorToInstantiatedTokenToDependencies =
+      new WeakMap<Injector|LView, WeakMap<Type<unknown>, InjectedService[]>>();
+  frameworkDIDebugData.instanceToInjector = new WeakMap<object, Injector>();
+  frameworkDIDebugData.injectorToProviders = new WeakMap<Injector|LView, ProviderRecord[]>();
+  frameworkDIDebugData.standaloneInjectorToComponent = new WeakMap<Injector, Type<unknown>>();
+
+  setInjectorProfiler(({data, type}: InjectorProfilerEvent) => {
+    if (type === InjectorProfilerEventType.Inject) {
+      const context = getDebugInjectContext();
+
+      if (context === undefined || context.token === null) {
+        return;
+      }
+
+      if (typeof context.token !== 'object' && typeof context.token !== 'function') {
+        return;
+      }
+
+      const {token, value, flags} = data as InjectedService;
+
+      let injectorKey: Injector|LView = context.injector;
+
+      if (context.injector instanceof DebugNodeInjector) {
+        injectorKey = context.injector.lView!;
+      }
+
+      if (!frameworkDIDebugData.injectorToInstantiatedTokenToDependencies.has(injectorKey)) {
+        frameworkDIDebugData.injectorToInstantiatedTokenToDependencies.set(
+            injectorKey, new WeakMap<Type<unknown>, InjectedService[]>());
+      }
+
+      if (!frameworkDIDebugData.injectorToInstantiatedTokenToDependencies.get(injectorKey)!.has(
+              context.token)) {
+        frameworkDIDebugData.injectorToInstantiatedTokenToDependencies.get(injectorKey)!.set(
+            context.token, []);
+      }
+
+      frameworkDIDebugData.injectorToInstantiatedTokenToDependencies.get(injectorKey)!
+          .get(context.token)!.push({token, value, flags});
```

Comment:
May be extract into a const for readability (const name is TBD, this is just to illustrate the idea)?
```suggestion
      const dependencyTracker = frameworkDIDebugData.injectorToInstantiatedTokenToDependencies;
      if (!dependencyTracker.has(injectorKey)) {
        // We see this key for the first time, create a WeakMap to collect deps.
        dependencyTracker.set(injectorKey, new WeakMap<Type<unknown>, InjectedService[]>());
      }

      if (!dependencyTracker.get(injectorKey)!.has(context.token)) {
        // This dependency is not yet in the WeakMap, add it.
        dependencyTracker.get(injectorKey)!.set(context.token, []);
      }

      // Register information about a token, its value and flags used for injection.
      dependencyTracker.get(injectorKey)!.get(context.token)!.push({token, value, flags});
```